src/ast/passes/semantic_analyser.cpp

```diff
@@ -2105,8 +2182,13 @@ void SemanticAnalyser::visit(Cast &cast)
   // case : BEGIN { @foo = (struct Foo)0; }
   // case : profile:hz:99 $task = (struct task_struct *)curtask.
   if (cast.type.GetAS() == AddrSpace::none) {
-    ProbeType type = single_provider_type();
-    cast.type.SetAS(find_addrspace(type));
+    if (auto probe = dynamic_cast<Probe *>(scope_)) {
+      ProbeType type = single_provider_type(probe);
+      cast.type.SetAS(find_addrspace(type));
+    } else {
+      // Assume kernel space for data in subprogs
```

Comment:
Yes, the issue is this probably requires a new language feature, since in the context of a subprog, there is no way to tell if you are accessing userspace or kernel space data based on pointers alone.